radicle/src/cob/identity.rs

```diff
@@ -0,0 +1,680 @@
+use std::{ops::Deref, str::FromStr};
+
+use crypto::{PublicKey, Signature};
+use once_cell::sync::Lazy;
+use radicle_cob::{ObjectId, TypeName};
+use radicle_crdt::{clock, GMap, Gate, LWWMap, LWWReg, Max, Redactable, Semilattice};
+use radicle_crypto::{Signer, Verified};
+use radicle_git_ext::Oid;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::{
+    cob::{
+        self,
+        common::Timestamp,
+        store::{self, FromHistory as _, Transaction},
+    },
+    identity::{doc::DocError, Identity},
+    prelude::Doc,
+    storage::{git as storage, RemoteId},
+};
+
+use super::{
+    thread::{self, Thread},
+    Author, OpId,
+};
+
+/// The logical clock we use to order operations to patches.
+pub use clock::Lamport as Clock;
+
+/// Type name of a patch.
+pub static TYPENAME: Lazy<TypeName> =
+    Lazy::new(|| FromStr::from_str("xyz.radicle.identity.proposal").expect("type name is valid"));
+
+pub type Op = cob::Op<Action>;
+
+pub type ProposalId = ObjectId;
+
+pub type RevisionId = OpId;
+
+/// Proposal operation.
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
+#[serde(tag = "type", rename_all = "camelCase")]
+pub enum Action {
+    Accept {
+        revision: RevisionId,
+        signature: Signature,
+    },
+    Edit {
+        title: String,
+        description: String,
+    },
+    Publish,
+    Redact {
+        revision: RevisionId,
+    },
+    Reject {
+        revision: RevisionId,
+    },
+    Revision {
+        proposed: Doc<Verified>,
+        previous: Identity<Oid>,
```

Comment:
Do we really need to include the whole identity, or would the `oid` suffice?

Reply:
So I explored this and it's quite interesting. The `Identity` is used for getting the difference between the proposed document and the previous one as well as checking if a quorum is reached. Now, if I change this field to `Oid` it means we need to pass the previous `Identity` to those calls. In turn, that means `previous` wasn't actually used in any other way. So in essence we're always proposing the change against the current `Identity`. 

This seems to make sense to me, it's kind of like always comparing to the latest canonical branch. I'm trying to think if this has any downsides -- one potential might be that a review is being performed on a current `Identity` that is out of date, however, correct me if I'm wrong, this wouldn't happen because they pulled the identity proposal and so they should have _at least_ the identity that the peer proposed, right?

Reply:
Yeah, what might happen though is that different users see different diffs, because their latest is different? I think having a `previous: Oid` (could be blob or commit id) might make sense to be able to throw an error in those cases.

So for example, I have an out of date id branch, I make a proposal, you receive it, and you see that it's a proposal for an old commit -- you reject it. Then I create a new up to date proposal after syncing.

Reply:
So, I've reworked this and now emit warnings when there's a mismatch. I also introduced the subcommand `rebase` that updates to the `Identity::current` in the next revision.